File: src/lib/matching.ts
import { embedTexts } from "@/lib/openai";
import { extractKeywords, inferJobDetails, tokenize } from "@/lib/keywords";
import type { MatchResult, ResumeRecord } from "@/lib/types";

export async function rankResumes(jd: string, resumes: ResumeRecord[]): Promise<MatchResult[]> {
  const jdKeywords = extractKeywords(jd, 45);
  const embeddings = await embedTexts([jd, ...resumes.map((resume) => resume.text)]).catch(() => null);
  const jdEmbedding = embeddings?.[0];
  const resumeEmbeddings = embeddings?.slice(1) ?? [];
  const jobDetails = inferJobDetails(jd);

  const matches = resumes.map((resume, index) => {
    const resumeKeywords = resume.keywords.length > 0 ? resume.keywords : extractKeywords(resume.text, 45);
    const resumeKeywordSet = new Set(resumeKeywords);
    const matchedKeywords = jdKeywords.filter((keyword) => resumeKeywordSet.has(keyword)).slice(0, 18);
    const missingKeywords = jdKeywords.filter((keyword) => !resumeKeywordSet.has(keyword)).slice(0, 12);
    const keywordScore = jdKeywords.length ? matchedKeywords.length / jdKeywords.length : 0;
    const skillsScore = getSkillScore(jdKeywords, resumeKeywords);
    const roleScore = getRoleScore(jobDetails.jobTitle, resume.text);
    const semanticScore =
      jdEmbedding && resumeEmbeddings[index]?.length
        ? cosineSimilarity(jdEmbedding, resumeEmbeddings[index])
        : lexicalSimilarity(jd, resume.text);

    const blendedScore = clamp01(
      semanticScore * 0.42 + keywordScore * 0.28 + skillsScore * 0.2 + roleScore * 0.1
    );

    return {
      resumeId: resume.id,
      resumeName: resume.originalName,
      score: Math.round(blendedScore * 100),
      skillsScore: Math.round(skillsScore * 100),
      keywordScore: Math.round(keywordScore * 100),
      roleScore: Math.round(roleScore * 100),
      semanticScore: Math.round(semanticScore * 100),
      matchedKeywords,
      missingKeywords,
      rationale: buildRationale(blendedScore, matchedKeywords, missingKeywords, resume.kind)
    } satisfies MatchResult;
  });

  return matches.sort((a, b) => b.score - a.score);
}

function getSkillScore(jdKeywords: string[], resumeKeywords: string[]) {
  const jdSkillTerms = jdKeywords.filter((keyword) => keyword.includes(".") || keyword.length > 3);
  if (jdSkillTerms.length === 0) {
    return 0;
  }

  const resumeSet = new Set(resumeKeywords);
  const hits = jdSkillTerms.filter((keyword) => resumeSet.has(keyword));
  return hits.length / jdSkillTerms.length;
}

function getRoleScore(jobTitle: string, resumeText: string) {
  const roleTokens = tokenize(jobTitle);
  if (roleTokens.length === 0) {
    return 0.5;
  }

  const resumeTokens = new Set(tokenize(resumeText));
  const hits = roleTokens.filter((token) => resumeTokens.has(token)).length;
  return hits / roleTokens.length;
}

function lexicalSimilarity(a: string, b: string) {
  const aTokens = tokenize(a);
  const bTokens = tokenize(b);
  const vectorA = termVector(aTokens);
  const vectorB = termVector(bTokens);
  const terms = new Set([...vectorA.keys(), ...vectorB.keys()]);
  let dot = 0;
  let magA = 0;
  let magB = 0;

  for (const term of terms) {
    const av = vectorA.get(term) ?? 0;
    const bv = vectorB.get(term) ?? 0;
    dot += av * bv;
    magA += av * av;
    magB += bv * bv;
  }

  if (magA === 0 || magB === 0) {
    return 0;
  }

  return dot / (Math.sqrt(magA) * Math.sqrt(magB));
}

function termVector(tokens: string[]) {
  const vector = new Map<string, number>();
  for (const token of tokens) {
    vector.set(token, (vector.get(token) ?? 0) + 1);
  }
  return vector;
}

function cosineSimilarity(a: number[], b: number[]) {
  let dot = 0;
  let magA = 0;
  let magB = 0;

  for (let index = 0; index < Math.min(a.length, b.length); index += 1) {
    dot += a[index] * b[index];
    magA += a[index] * a[index];
    magB += b[index] * b[index];
  }

  if (magA === 0 || magB === 0) {
    return 0;
  }

  return clamp01(dot / (Math.sqrt(magA) * Math.sqrt(magB)));
}

function buildRationale(score: number, matchedKeywords: string[], missingKeywords: string[], kind: string) {
  if (score >= 0.75) {
    return `Strong ${kind} resume match with ${matchedKeywords.length} JD keywords already present.`;
  }

  if (score >= 0.55) {
    return `Usable match. Optimization should add ${missingKeywords.slice(0, 3).join(", ") || "targeted keywords"}.`;
  }

  return `Lower fit. The JD calls for ${missingKeywords.slice(0, 4).join(", ") || "skills not visible in this resume"}.`;
}

function clamp01(value: number) {
  return Math.max(0, Math.min(1, value));
}

File: app/matching.py
from __future__ import annotations

import math
from collections import Counter
from typing import Any

from app.ai import embed_texts
from app.keywords import extract_keywords, infer_job_details, tokenize


def rank_resumes(jd: str, resumes: list[dict[str, Any]]) -> list[dict[str, Any]]:
    jd_keywords = extract_keywords(jd, 45)
    embeddings = None
    try:
        embeddings = embed_texts([jd] + [resume["text"] for resume in resumes])
    except Exception:
        embeddings = None

    jd_embedding = embeddings[0] if embeddings else None
    resume_embeddings = embeddings[1:] if embeddings else []
    details = infer_job_details(jd)

    matches: list[dict[str, Any]] = []
    for index, resume in enumerate(resumes):
        resume_keywords = resume.get("keywords") or extract_keywords(resume["text"], 45)
        resume_keyword_set = set(resume_keywords)
        matched = [keyword for keyword in jd_keywords if keyword in resume_keyword_set][:18]
        missing = [keyword for keyword in jd_keywords if keyword not in resume_keyword_set][:12]

        keyword_score = len(matched) / len(jd_keywords) if jd_keywords else 0
        skills_score = skill_score(jd_keywords, resume_keywords)
        role_score_value = role_score(details["job_title"], resume["text"])
        semantic_score = (
            cosine(jd_embedding, resume_embeddings[index])
            if jd_embedding and index < len(resume_embeddings)
            else lexical_similarity(jd, resume["text"])
        )

        blended = clamp(semantic_score * 0.42 + keyword_score * 0.28 + skills_score * 0.2 + role_score_value * 0.1)
        matches.append(
            {
                "resume_id": resume["id"],
                "resume_name": resume["original_name"],
                "score": round(blended * 100),
                "skills_score": round(skills_score * 100),
                "keyword_score": round(keyword_score * 100),
                "role_score": round(role_score_value * 100),
                "semantic_score": round(semantic_score * 100),
                "matched_keywords": matched,
                "missing_keywords": missing,
                "rationale": rationale(blended, matched, missing, resume["kind"]),
            }
        )

    return sorted(matches, key=lambda item: item["score"], reverse=True)


def skill_score(jd_keywords: list[str], resume_keywords: list[str]) -> float:
    skill_terms = [keyword for keyword in jd_keywords if "." in keyword or len(keyword) > 3]
    if not skill_terms:
        return 0
    resume_set = set(resume_keywords)
    return len([keyword for keyword in skill_terms if keyword in resume_set]) / len(skill_terms)


def role_score(job_title: str, resume_text: str) -> float:
    role_tokens = tokenize(job_title)
    if not role_tokens:
        return 0.5
    resume_tokens = set(tokenize(resume_text))
    return len([token for token in role_tokens if token in resume_tokens]) / len(role_tokens)


def lexical_similarity(a: str, b: str) -> float:
    vector_a = Counter(tokenize(a))
    vector_b = Counter(tokenize(b))
    terms = set(vector_a) | set(vector_b)
    dot = sum(vector_a[term] * vector_b[term] for term in terms)
    mag_a = math.sqrt(sum(value * value for value in vector_a.values()))
    mag_b = math.sqrt(sum(value * value for value in vector_b.values()))
    return dot / (mag_a * mag_b) if mag_a and mag_b else 0


def cosine(a: list[float], b: list[float]) -> float:
    if not a or not b:
        return 0
    dot = sum(left * right for left, right in zip(a, b))
    mag_a = math.sqrt(sum(value * value for value in a))
    mag_b = math.sqrt(sum(value * value for value in b))
    return clamp(dot / (mag_a * mag_b)) if mag_a and mag_b else 0


def rationale(score: float, matched: list[str], missing: list[str], kind: str) -> str:
    if score >= 0.75:
        return f"Strong {kind} resume match with {len(matched)} JD keywords already visible."
    if score >= 0.55:
        targets = ", ".join(missing[:3]) or "targeted keywords"
        return f"Usable match. Optimization should emphasize {targets}."
    targets = ", ".join(missing[:4]) or "skills not visible in this resume"
    return f"Lower fit. The JD calls for {targets}."


def clamp(value: float) -> float:
    return max(0.0, min(1.0, value))
